File: ios/FinanceFlowKit/Sources/FinanceFlowKit/Graph/Path.swift
import Foundation

public enum Path {
    /// The single straight-line path through the graph given current answers.
    /// Unanswered decisions optimistically follow their "yes" branch.
    /// Direct port of `linearPath` in `src/graph/path.ts`.
    public static func linearPath(_ state: AppState) -> [NodeId] {
        var out: [NodeId] = []
        var visited: Set<NodeId> = []
        var cur: NodeId? = .Start

        while let node = cur {
            if visited.contains(node) { break }
            visited.insert(node)
            out.append(node)
            let g = Flowchart.node(node)
            if g.kind == .decision {
                let answer = state.decisions[g.decisionId!]
                if let answer {
                    cur = g.edges.first(where: { $0.when == answer })?.to
                } else {
                    cur = g.edges.first(where: { $0.when == .yes })?.to ?? g.edges.first?.to
                }
            } else {
                cur = g.edges.first?.to
            }
        }
        return out
    }

    /// Previous / next node along the linear path (with a topology fallback for
    /// off-path nodes). Direct port of `neighbors` in `src/graph/path.ts`.
    public static func neighbors(_ state: AppState, of id: NodeId) -> (prev: NodeId?, next: NodeId?) {
        let path = linearPath(state)
        if let idx = path.firstIndex(of: id) {
            return (
                prev: idx > 0 ? path[idx - 1] : nil,
                next: idx < path.count - 1 ? path[idx + 1] : nil
            )
        }

        // Off-path fallback: graph topology only.
        let g = Flowchart.node(id)
        let next: NodeId?
        if g.kind == .decision {
            let answer = state.decisions[g.decisionId!] ?? .yes
            next = g.edges.first(where: { $0.when == answer })?.to ?? g.edges.first?.to
        } else {
            next = g.edges.first?.to
        }
        let prev = Flowchart.graph.first(where: { $0.edges.contains(where: { $0.to == id }) })?.id
        return (prev, next)
    }
}

File: src/graph/path.ts
import type { AppState, NodeId } from "../state/schema";
import { GRAPH, GRAPH_BY_ID, type GraphNode } from "./flowchart";

export function linearPath(state: Pick<AppState, "decisions">): NodeId[] {
  const out: NodeId[] = [];
  const visited = new Set<NodeId>();
  let cur: NodeId | null = "Start";
  while (cur) {
    if (visited.has(cur)) break;
    visited.add(cur);
    out.push(cur);
    const node: GraphNode = GRAPH_BY_ID[cur];
    if (node.kind === "decision") {
      const ans = state.decisions[node.decisionId!];
      if (ans) {
        cur = node.edges.find((e) => e.when === ans)?.to ?? null;
      } else {
        cur = node.edges.find((e) => e.when === "yes")?.to ?? node.edges[0]?.to ?? null;
      }
    } else {
      cur = node.edges[0]?.to ?? null;
    }
  }
  return out;
}

export function neighbors(
  state: Pick<AppState, "decisions">,
  id: NodeId,
): { prev: NodeId | null; next: NodeId | null } {
  const path = linearPath(state);
  const idx = path.indexOf(id);
  if (idx >= 0) {
    return {
      prev: idx > 0 ? path[idx - 1] : null,
      next: idx < path.length - 1 ? path[idx + 1] : null,
    };
  }

  // Off-path fallback: graph topology only
  const node: GraphNode = GRAPH_BY_ID[id];
  let next: NodeId | null = null;
  if (node.kind === "decision") {
    const ans = state.decisions[node.decisionId!];
    next = node.edges.find((e) => e.when === (ans ?? "yes"))?.to ?? node.edges[0]?.to ?? null;
  } else {
    next = node.edges[0]?.to ?? null;
  }
  let prev: NodeId | null = null;
  for (const n of GRAPH) {
    if (n.edges.some((e) => e.to === id)) {
      prev = n.id;
      break;
    }
  }
  return { prev, next };
}
